src/sparse6.rs

```diff
@@ -0,0 +1,269 @@
+use pyo3::prelude::*;
+use pyo3::types::PyAny;
+use crate::graph6::IOError;
+use crate::graph::PyGraph;
+use petgraph::graph::NodeIndex;
+use petgraph::prelude::Undirected;
+use crate::StablePyGraph;
+use std::iter;
+
+fn parse_n(bytes: &[u8], pos: usize) -> Result<(usize, usize), IOError> {
+    if pos >= bytes.len() {
+        return Err(IOError::NonCanonicalEncoding);
+    }
+    let first = bytes[pos];
+    if first < 63 || first > 126 {
+        return Err(IOError::InvalidSizeChar);
+    }
+    if first != 126 {
+        return Ok(((first - 63) as usize, pos + 1));
+    }
+    // first == 126 -> extended form. Look ahead without advancing permanently.
+    if pos + 1 >= bytes.len() {
+        return Err(IOError::NonCanonicalEncoding);
+    }
+    let second = bytes[pos + 1];
+    if second == 126 {
+        // 8 byte form: 126 126 R(x) where R(x) is 36 bits -> 6 bytes
+        if bytes.len() < pos + 2 + 6 {
+            return Err(IOError::NonCanonicalEncoding);
+        }
+        let mut val: u64 = 0;
+        for i in 0..6 {
+            let b = bytes[pos + 2 + i];
+            if b < 63 || b > 126 {
+                return Err(IOError::InvalidSizeChar);
+            }
+            val = (val << 6) | ((b - 63) as u64);
+        }
+        return Ok((val as usize, pos + 2 + 6));
+    } else {
+        // 4 byte form: 126 R(x) where R(x) is 18 bits -> 3 bytes
+        if bytes.len() < pos + 1 + 3 {
+            return Err(IOError::NonCanonicalEncoding);
+        }
+        let mut val: usize = 0;
+        for i in 0..3 {
+            let b = bytes[pos + 1 + i];
+            if b < 63 || b > 126 {
+                return Err(IOError::InvalidSizeChar);
+            }
+            val = (val << 6) | ((b - 63) as usize);
+        }
+        if val == 64032 {
+            eprintln!("DEBUG sparse6 parse_n anomaly: pos={} bytes_prefix={:?} triple={:?}", pos, &bytes[..std::cmp::min(bytes.len(),10)], [&bytes[pos+1], &bytes[pos+2], &bytes[pos+3]]);
+        }
+        return Ok((val, pos + 1 + 3));
+    }
+}
+
+fn bits_from_bytes(bytes: &[u8], start: usize) -> Result<Vec<u8>, IOError> {
+    let mut bits = Vec::new();
+    for &b in bytes.iter().skip(start) {
```

Comment:
`for &byte in &bytes[start..]`? or even just slice in the caller, this function is only used once